packages/timeline/src/browser/timeline-widget.tsx

```diff
@@ -0,0 +1,208 @@
+        const current = this.applicationShell.currentWidget;
+        const currentUri = NavigatableWidget.is(current) ? current.getResourceUri() : undefined;
+        if (event.removed) {
+            for (const source of event.removed) {
+                this.timelinesBySource.delete(source);
+            }
+
+            if (currentUri) {
+                this.loadTimeline(currentUri, true);
+            }
+        } else if (event.added) {
+            if (currentUri) {
+                event.added.forEach( source => this.loadTimelineForSource(source, currentUri, true));
+            }
+        }
+    }
+
+    async loadTimelineForSource(source: string, uri: URI, reset: boolean): Promise<void> {
+        if (reset) {
+            this.timelinesBySource.delete(source);
+        }
+        let timeline = this.timelinesBySource.get(source);
+        const cursor = timeline?.cursor;
+        const options = { cursor: reset ? undefined : cursor, limit: TimelineTreeWidget.PAGE_SIZE };
+        const timelineResult = await this.timelineService.getTimeline(source, uri, options);
+        if (timelineResult) {
+            const items = timelineResult.items;
+            if (items) {
+                if (timeline) {
+                    timeline.add(items);
+                    timeline.cursor = timelineResult.paging?.cursor;
+                } else {
+                    timeline = new TimelineAggregate(timelineResult);
+                }
+                this.timelinesBySource.set(source, timeline);
+                this.resourceWidget.model.updateTree(timeline.items, !!timeline.cursor);
+            }
+        }
+    }
+
+    async loadTimeline(uri: URI, reset: boolean): Promise<void> {
+        for (const source of this.timelineService.getSources().map(s => s.id)) {
+            this.loadTimelineForSource(source, uri, reset);
+        }
+    }
+
+    refreshList(): void {
+        const uri = this.getCurrentWidgetUri();
+        if (uri) {
+            this.timelineEmptyWidget.hide();
+            this.resourceWidget.show();
+            this.loadTimeline(uri, true);
+        } else if (!this.suitableWidgetsOpened()) {
+            this.timelineEmptyWidget.show();
+            this.resourceWidget.hide();
+        }
+    }
+
+    private suitableWidgetsOpened(): boolean {
+        return !!toArray(this.applicationShell.mainPanel.widgets()).find(widget => {
+            if (NavigatableWidget.is(widget)) {
+                const uri = widget.getResourceUri();
+                if (uri?.scheme && this.timelineService.getSchemas().indexOf(uri?.scheme) > -1) {
+                    return true;
+                }
+            }
+        });
+    }
+
+    private getCurrentWidgetUri(): URI | undefined {
+        const current = this.applicationShell.currentWidget;
+        return  NavigatableWidget.is(current) ? current.getResourceUri() : undefined;
+    }
+
+    protected get containerLayout(): PanelLayout {
```

Comment:
`PanelLayout | undefined`, if widget is disposed, underlying layout will be null, please adjust clients to respect it

Reply:
done